### PR TITLE
Add support for SM2235 and SM2335 LED drivers

### PR DESCRIPTION
**Pull request by Cossid**
# What does this implement/fix?

This implements a base LED driver for Sunmoon Microelectronics 10 bit LED driver base support the SM2235 and SM2335 LED drivers.  The SM2335 LED driver can be found in the SwitchBot Color Bulb.

Note: The existing SM2135 driver is 8 bit and substantially different, so must remain a separate component not using this new base.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2514

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

SM2335:
```yaml
    # Example configuration entry
    sm2335:
      data_pin: GPIO4
      clock_pin: GPIO5
      max_power_color_channels: 9
      max_power_white_channels: 9

    # Individual outputs
    output:
      - platform: sm2335
        id: output_red
        channel: 1
      - platform: sm2335
        id: output_green
        channel: 0
      - platform: sm2335
        id: output_blue
        channel: 2
      - platform: sm2335
        id: output_coldwhite
        channel: 4
      - platform: sm2335
        id: output_warmwhite
        channel: 3
```

SM2235:
```yaml
    # Example configuration entry
    sm2235:
      data_pin: GPIO4
      clock_pin: GPIO5
      max_power_color_channels: 9
      max_power_white_channels: 9

    # Individual outputs
    output:
      - platform: sm2235
        id: output_red
        channel: 1
      - platform: sm2235
        id: output_green
        channel: 0
      - platform: sm2235
        id: output_blue
        channel: 2
      - platform: sm2235
        id: output_coldwhite
        channel: 4
      - platform: sm2235
        id: output_warmwhite
        channel: 3
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Cossid** [comment]
To the best of my understanding, all remaining CI failures are unrelated to changes made by this PR.

**Comment by sebcaps** [comment]
@Cossid : https://github.com/esphome/issues/issues/3423 ?

**Comment by ssieb** [comment]
Other than the trivial setters, move the function contents to the `.cpp` file.

**Comment by Cossid** [comment]
> Other than the trivial setters, move the function contents to the `.cpp` file.

Yeah, I started with SM2135 as a base, but that does seem poorly written.  SM2135 needs some work too, but it will require breaking changes.  I will likely address that driver in the future as I work on a few other LED drivers.

@ssieb I have adjusted this driver to closer emulate the my9231 driver header/source structure, please re-review that this meets expectations.

**Comment by Cossid** [comment]
Just adding a note that there was a request for a SM2235 driver on discord, which looks like it shares the majority of code with the 2335 driver.  I have created a fork branch named sm_1024bit_base which shares as much code as I was able to, and pending feedback from users with sm2235 drivers, this PR may end up abandoned in favor of that one which supports both SM2235 and SM2335.  If this one is merged, it should be a non-breaking change to the other branch.

**Comment by Cossid** [comment]
~~Last remaining CI failure is due to a naming convention (using an underscore in a class name), however, this appears to be the norm for other classes used as a base class.  I am leaving it as is for now to match existing code, but can change if needed.~~

After further review, the underscore is only used in the component directory name and namespace, renamed class to match linter expectations.

**Comment by ryanjohnsontv** [comment]
Just wanted to add that I used this branch to compile firmware for my Switchbot RGB+CT E26 W1401400 bulb (ESP32-C3), and it works flawlessly. Thank you so much for your work!

File: esphome/components/sm10bit_base/__init__.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            min=0, max=15
+        ),
+        cv.Optional(CONF_MAX_POWER_WHITE_CHANNELS, default=4): cv.int_range(
+            min=0, max=15
+        ),
+    }
+).extend(cv.COMPONENT_SCHEMA)
+
+
+@coroutine
```
Comment (jesserockz) [inline esphome/components/sm10bit_base/__init__.py]: Not needed
```suggestion
```

File: esphome/components/sm10bit_base/__init__.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import pins
+from esphome.const import (
+    CONF_CLOCK_PIN,
+    CONF_DATA_PIN,
+    CONF_ID,
+)
+from esphome.core import coroutine
```
Comment (jesserockz) [inline esphome/components/sm10bit_base/__init__.py]:
```suggestion
```

**Review by jesserockz — APPROVED** [review]
Minor change, but otherwise good to go

File: esphome/components/sm10bit_base/__init__.py (hunk 1 above)
Comment (Cossid) [reply esphome/components/sm10bit_base/__init__.py]: Yeah, I didn't fully understand how @coroutine worked, it was in other _base or led drivers, so I copied.

**Pull request merged**